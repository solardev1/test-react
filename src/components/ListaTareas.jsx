import React, { useState, useEffect } from 'react';
import { useNavigate } from "react-router-dom";

const ListaTareas = () => {
  const [tareas, setTareas] = useState(() => {
    const storedTareas = localStorage.getItem('tareas');
    return storedTareas ? JSON.parse(storedTareas) : [];
  });
  const [tituloTarea, setTituloTarea] = useState('');
  const [descripcionTarea, setDescripcionTarea] = useState('');
  const [prioridadTarea, setPrioridadTarea] = useState('Alta');
  const [filtro, setFiltro] = useState('');
  const [tareasFiltradas, setTareasFiltradas] = useState([]);
  const navigate = useNavigate();


  useEffect(() => {
    const usuario = JSON.parse(localStorage.getItem('usuario'));
    if (!usuario || !usuario.logueado) {
      navigate('/login');
    }
  }, []);
  
  useEffect(() => {
    localStorage.setItem('tareas', JSON.stringify(tareas));
  }, [tareas]);

  useEffect(() => {
    if (filtro === '') {
      setTareasFiltradas(tareas);
    } else {
      const nuevasTareasFiltradas = tareas.filter(tarea =>
        tarea.titulo.toLowerCase().includes(filtro.toLowerCase()) ||
        tarea.descripcion.toLowerCase().includes(filtro.toLowerCase()) ||
        tarea.prioridad.toLowerCase().includes(filtro.toLowerCase())
      );
      setTareasFiltradas(nuevasTareasFiltradas);
    }
  }, [tareas, filtro]);

  const handleTituloTareaChange = (event) => {
    setTituloTarea(event.target.value);
  };

  const handleDescripcionTareaChange = (event) => {
    setDescripcionTarea(event.target.value);
  };

  const handlePrioridadTareaChange = (event) => {
    setPrioridadTarea(event.target.value);
  };

  const handleAgregarTarea = () => {
    if (tituloTarea.trim() !== '' && descripcionTarea.trim() !== '') {
      const nuevaTarea = {
        titulo: tituloTarea,
        descripcion: descripcionTarea,
        prioridad: prioridadTarea,
        completada: false
      };
      setTareas([...tareas, nuevaTarea]);
      setTituloTarea('');
      setDescripcionTarea('');
      setPrioridadTarea('Alta');
    }
  };

  const handleFiltroChange = (event) => {
    setFiltro(event.target.value);
  };
  const handleCompletarTarea = (index) => {
    const nuevasTareas = [...tareas];
    nuevasTareas[index].completada = !nuevasTareas[index].completada;
    setTareas(nuevasTareas);
  };

  const handleEliminarTarea = (index) => {
    const nuevasTareas = [...tareas];
    nuevasTareas.splice(index, 1);
    setTareas(nuevasTareas);
  };

  return (
    <div className="container mt-5">
      <h2 className="text-center mb-4">Lista de Tareas</h2>
      <div className="mb-3">
        <input
          type="text"
          className="form-control"
          placeholder="Buscar por nombre, descripción o prioridad"
          value={filtro}
          onChange={handleFiltroChange}
        />
      </div>
      <div className="row mb-3">
        <div className="col-md-4">
          <div className="form-group">
            <label htmlFor="tituloTarea">Título de Tarea:</label>
            <input
              type="text"
              className="form-control"
              id="tituloTarea"
              value={tituloTarea}
              onChange={handleTituloTareaChange}
            />
          </div>
        </div>
        <div className="col-md-4">
          <div className="form-group">
            <label htmlFor="descripcionTarea">Descripción:</label>
            <input
              type="text"
              className="form-control"
              id="descripcionTarea"
              value={descripcionTarea}
              onChange={handleDescripcionTareaChange}
            />
          </div>
        </div>
        <div className="col-md-4">
          <div className="form-group">
            <label htmlFor="prioridadTarea">Prioridad:</label>
            <select
              className="form-control"
              id="prioridadTarea"
              value={prioridadTarea}
              onChange={handlePrioridadTareaChange}
            >
              <option value="Alta">Alta</option>
              <option value="Media">Media</option>
              <option value="Baja">Baja</option>
            </select>
          </div>
        </div>
      </div>
      <button className="btn btn-primary mb-3" onClick={handleAgregarTarea}>
        Agregar Tarea
      </button>
      <table className="table">
        <thead>
          <tr>
            <th>Título</th>
            <th>Descrip.</th>
            <th>Prioridad</th>
            <th>Hecho</th>
          </tr>
        </thead>
        <tbody>
          {tareasFiltradas.map((tarea, index) => (
            <tr key={index}>
              <td>{tarea.titulo}</td>
              <td>{tarea.descripcion}</td>
              <td>{tarea.prioridad}</td>
              <td>
                <input
                  type="checkbox"
                  checked={tarea.completada}
                  onChange={() => handleCompletarTarea(index)}
                />
              </td>
              <td>
                <button
                  className="btn btn-danger btn-sm"
                  onClick={() => handleEliminarTarea(index)}
                >
                  Eliminar
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default ListaTareas;
