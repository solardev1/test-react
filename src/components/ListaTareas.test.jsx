import React from 'react';
import { shallow } from 'enzyme';
import ListaTareas from './ListaTareas';
import { useNavigate } from 'react-router-dom';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

describe('ListaTareas', () => {
  let wrapper;

  beforeEach(() => {
    // Simulamos que el usuario no está autenticado
    jest.spyOn(React, 'useEffect').mockImplementationOnce(f => f());
    jest.spyOn(window.localStorage.__proto__, 'getItem').mockReturnValueOnce(JSON.stringify({ logueado: false }));
    wrapper = shallow(<ListaTareas />);
  });

  it('debería redirigir a la página de inicio de sesión si el usuario no está autenticado', () => {
    expect(useNavigate).toHaveBeenCalled();
  });

  it('debería renderizar correctamente si el usuario está autenticado', () => {
    // Simulamos que el usuario está autenticado
    const usuario = { logueado: true }; // Puedes agregar otros atributos según sea necesario
    jest.spyOn(window.localStorage.__proto__, 'getItem').mockReturnValueOnce(JSON.stringify(usuario));
    wrapper = shallow(<ListaTareas />);
    
    // Aquí puedes agregar las expectativas para verificar que el componente se renderiza correctamente
  });
});
