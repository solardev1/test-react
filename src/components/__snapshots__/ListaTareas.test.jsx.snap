// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListaTareas Component renders correctly 1`] = `ShallowWrapper {}`;
